### DISCUSSION TITLE
Data assets / pipelines that are collections

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#12061] [Q&A]
### Use case: collections of data pipelines with identical structure

Some Dagster use cases look like collections of data pipelines with identical structure, but different data.

For example:

- You are a business that trains an ML model for each one of your customers, using data provided by that customer. The training process is the same across all customers, but the data is different. You want to be able to launch a run that (re)trains the model for a particular customer, without overwriting the models for other customers.
- You are a research lab that produces a set of analyses from each experiment that you run. The analysis functions are identical, but the experimental data is different.  You want to be able to launch a run that derives the analyses for a particular experiment, without overwriting the analyses for other experiments.

### Partitioned assets to the rescue

We recommend modeling these kinds of use cases as **graphs of partitioned assets**. A partitioned asset represents a collection of objects that are all computed the same way.  A set of partitioned assets can share the same `PartitionsDefinition`.

By default, Dagster will assume that, if asset A and B have the same partitioning, and B depends on A, then partition X of B is derived from partition X of A. This can be customized by using a non-default `PartitionMapping`. 

Using a partitioned asset instead of many individual assets has a few advantages, by exploiting the fact that the pipelines have identical structure:

- It allows Dagster to represent the asset graph in a more visually-condensed way in the web UI.
- It allows Dagster to be more performant.
- It allows you to run ************backfills,************ which allow you to materialize the assets across all your partitions, as a single operation

### Example
Here’s a code example that corresponds to the example in the first bullet point of the first section:

```python
import pandas as pd
import pickle
from dagster import StaticPartitionsDefinition


customers_partitions_def = StaticPartitionsDefinition(["customer1", "customer2"])

@asset(partitions_def=customer_partitions_def)
def features(context) -> None:
    """Represents a dataset of features, per customer"""
    customer_name = context.partition_key
    customer_data = pd.read_csv(f"{customer_name}/source_data.csv")
    extracted_features = ...
    extracted_features.to_csv(f"{customer_name}/features.csv")

@asset(partitions_def=customer_partitions_def, deps=[features])
def ml_model(context) -> None:
    """Represents an ML model per customer"""
    customer_name = context.partition_key
    features = pd.read_csv(f"{customer_name}/features.csv")
    trained_model = ...
    with open(f"{customer_name}/model.pkl")
        pickle.dump(trained_model, f)
```

The above example handles I/O inside the body of the `@asset`-decorated function, but you can alternatively use an `IOManager`. The default IO manager will write outputs to a path that ends with `asset_key/partition_key`.

**Comment by yan91083** [comment]
Thanks for sharing this use case.  I wonder if there's a way you can pass what partitions user chose from upstream asset to downstream asset. For example, if user chose "customer1" while materializing **features**,  then the argument can pass to **ml_model**

**Reply by sryza** [reply]
Are you able to explain the scenario you're imagining in more detail?

If you're materializing both at the same time, you only need to specify the single partition that they're both executing on, e.g. "customer1".

If you're executing `ml_model` at a different time than you executed `features`, then `features` may have been executed in the past multiple different times to materialize the features for different customer, so you'd need to explicitly choose which of those customers you want to train the ML model for.
